File: Algorithms/Implementations/the_grid_search.c
#include <stdio.h>
#include <stdbool.h>

void print_pattern(int row_count, int col_count, char pattern[row_count][col_count])
{
    for (int i = 0; i < row_count; i++)
    {
        for (int j = 0; j < col_count; j++)
        {
            printf("%c", pattern[i][j]);
        }
        printf("\n");
    }
}

bool find_pattern(int source_row_count, int source_col_count, char source[source_row_count][source_col_count], int search_row_count, int search_col_count, char search[search_row_count][search_col_count])
{
    bool found;
    for (int i = 0; i < source_row_count - search_row_count + 1; i++)
    {
        for (int j = 0; j < source_col_count - search_col_count + 1; j++)
        {
            found = true;
            /* search for all 4 corners first, if matches, continue */
            if (source[i][j] == search[0][0] && source[i+search_row_count-1][j] == search[search_row_count-1][0] && source[i][j+search_col_count-1] == search[0][search_col_count-1] && source[i+search_row_count-1][j+search_col_count-1] == search[search_row_count-1][search_col_count-1])
            {
               for (int m = 0; m < search_row_count; m++)
                {
                    for (int n = 0; n < search_col_count; n++)
                    {
                        if (source[i+m][j+n] != search[m][n])
                        {
                            found = false;
                        }
                    }
                } 
            } else {
                found = false;
            }
            
            if (found == true)
            {
                return true;
            }
        }
    }
    return false;
}


bool find_pattern_brute_force(int source_row_count, int source_col_count, char source[source_row_count][source_col_count], int search_row_count, int search_col_count, char search[search_row_count][search_col_count])
{
    bool found;
    for (int i = 0; i < source_row_count - search_row_count + 1; i++)
    {
        for (int j = 0; j < source_col_count - search_col_count + 1; j++)
        {
            found = true;
            for (int m = 0; m < search_row_count; m++)
            {
                for (int n = 0; n < search_col_count; n++)
                {
                    if (source[i+m][j+n] != search[m][n])
                    {
                        found = false;
                    }
                }
            }
            if (found == true)
            {
                return true;
            }
        }
    }
    return false;
}

int main()
{
    int test_count;
    scanf("%d", &test_count);

    for (int i = 0; i < test_count; i ++)
    {
        int source_row, source_column, search_row, search_column;
        scanf("%d %d", &source_row, &source_column);
        
        char source[source_row][source_column];
        for(int j = 0; j < source_row; j++)
        {
            scanf("%s", source[j]);
        }
        
        scanf("%d %d", &search_row, &search_column);

        char search[search_row][search_column];
        for (int m = 0; m < search_row; m++)
        {
            scanf("%s", search[m]);
        }

        bool found;
        found = find_pattern(source_row, source_column, source, search_row, search_column, search);

        if (found == true)
        {
            printf("YES\n");
        } else {
            printf("NO\n");
        }
    }
    return 0;
}

File: Algorithms/Implementations/the_grid_search.py
def find_pattern(source, search):
    search_pattern_length = len(search[0])
    for source_row_number in xrange(len(source) - len(search) + 1):
        rows_found = 0
        column_start = 0
        try:
            while column_start < len(source[source_row_number]):
                column_start += source[source_row_number][column_start:].index(search[0])
                for search_row_number in xrange(1, len(search)):
                    try:
                        search_row_found = source[source_row_number + search_row_number][column_start:column_start+search_pattern_length].index(search[search_row_number])
                        rows_found +=1
                    except ValueError:
                        continue
                    except IndexError:
                        continue
                if rows_found == len(search) - 1:
                    return True
                else:
                    column_start = column_start + search_pattern_length
        except ValueError:
            continue

    return False

if __name__ == "__main__":
    test_count = int(raw_input().strip())
    for i in xrange(test_count):
        source_row, source_column = map(int, raw_input().strip().split())

        source = []
        for j in xrange(source_row):
            source.append(raw_input().strip())

        search_row, search_column = map(int, raw_input().strip().split())

        search = []
        for k in xrange(search_row):
            search.append(raw_input().strip())

        found = find_pattern(source, search)
        if found == True:
            print "YES"
        else:
            print "NO"
